### ISSUE TITLE
datadog-agent system-probe in CrashLoopBackoff

### ISSUE THREAD

**Issue by tantran-earnin** [bottlerocket-os/bottlerocket#3691]
**Image I'm using:**
Bottlerocket - 1.16.1 (aws-k8s-1.28)
Bottlerocket - 1.17.0 (aws-k8s-1.28)
Datadog helm chart - 3.49.7

**What I expected to happen:**
All Datadog containers running

**What actually happened:**
system-probe container crashing causing pod to be out of service

**How to reproduce the problem:**
Problem occurs when upgrading EKS node to Bottlerocket 1.17.0, problem disappears when pinning back down to 1.16.1.

```
k logs datadog-4c9fr -n redacted -c system-probe
2023-12-18 18:44:24 UTC | SYS-PROBE | WARN | (pkg/util/log/log.go:618 in func1) | Unknown key in config file: runtime_security_config.network.enabled
2023-12-18 18:44:24 UTC | SYS-PROBE | WARN | (pkg/util/log/log.go:618 in func1) | Unknown key in config file: runtime_security_config.activity_dump.path_merge.enabled
2023-12-18 18:44:24 UTC | SYS-PROBE | WARN | (pkg/util/log/log.go:618 in func1) | Unknown key in config file: runtime_security_config.syscall_monitor.enabled
2023-12-18 18:44:24 UTC | SYS-PROBE | WARN | (pkg/util/log/log.go:618 in func1) | Unknown key in config file: runtime_security_config.activity_dump.cgroup_wait_list_size
2023-12-18 18:44:24 UTC | SYS-PROBE | WARN | (pkg/util/log/log.go:618 in func1) | Agent configuration relax permissions constraint on the secret backend cmd, Group can read and exec
2023-12-18 18:44:24 UTC | SYS-PROBE | WARN | (pkg/util/log/log.go:613 in func1) | configuration key `runtime_security_config.activity_dump.cgroup_dump_timeout` is deprecated, use `runtime_security_config.activity_dump.dump_duration` instead
2023-12-18 18:44:24 UTC | SYS-PROBE | WARN | (pkg/security/probe/config/config.go:247 in getBool) | runtime_security_config.network.enabled has been deprecated: please set event_monitoring_config.network.enabled instead
2023-12-18 18:44:24 UTC | SYS-PROBE | WARN | (pkg/security/probe/probe_linux.go:1423 in NewProbe) | the current environment may be misconfigured: 3 errors occurred:
	* /etc/passwd doesn't seem to be a mountpoint
	* /etc/group doesn't seem to be a mountpoint
	* /host/sys/kernel/security doesn't seem to be a mountpoint
2023-12-18 18:44:48 UTC | SYS-PROBE | ERROR | (cmd/system-probe/api/module/loader.go:78 in Register) | error registering HTTP endpoints for module event_monitor: probes activation validation failed: 1 error occurred:
	* OneOf requirement failed, none of the following probes are running [{UID:security EBPFFuncName:kprobe_nf_nat_manip_pkt EBPFSection:}: couldn't enable kprobe {UID:security EBPFFuncName:kprobe_nf_nat_manip_pkt EBPFSection:}: cannot write "p:p_nf_nat_manip_pkt_security_349821 nf_nat_manip_pkt\n" to kprobe_events: write /sys/kernel/debug/tracing/kprobe_events: no such file or directory | {UID:security EBPFFuncName:kprobe_nf_nat_packet EBPFSection:}: couldn't enable kprobe {UID:security EBPFFuncName:kprobe_nf_nat_packet EBPFSection:}: cannot write "p:p_nf_nat_packet_security_349821 nf_nat_packet\n" to kprobe_events: write /sys/kernel/debug/tracing/kprobe_events: no such file or directory]

2023-12-18 18:44:48 UTC | SYS-PROBE | CRITICAL | (cmd/system-probe/subcommands/run/command.go:241 in startSystemProbe) | error while starting api server, exiting: failed to create system probe: no module could be loaded
Error: error while starting api server, exiting: failed to create system probe: no module could be loaded
```

**Comment by cbgbt** [comment]
Thanks for raising this issue, I'll take a look. Is this something that consistently occurs with each attempt to run the pod, or does it seem like a transient issue?

**Comment by tantran-earnin** [comment]
We utilize Bottlerocket OS in multiple clusters, and they all occurred/failed the same way constantly until we rolled back the image.

**Comment by cbgbt** [comment]
Thanks. The underlying error in the logs you provided seems to be:

```
cannot write "p:p_nf_nat_packet_security_349821 nf_nat_packet\n" to kprobe_events: write /sys/kernel/debug/tracing/kprobe_events: no such file or directory
```

I'll see if I can't work out why it can't seem to access that part of sysfs as of 1.17.0.

**Comment by cbgbt** [comment]
I see [here](https://github.com/DataDog/helm-charts/blob/main/charts/datadog/templates/_container-system-probe.yaml#L42C13-L42C20) where the probe chart mounts the kernel's debugfs into the container as `rw`.

I logged into the aws-k8s-1.28 variant using 1.16.1 and 1.17.0 and tried to set similar kprobes manually, using sheltie from the admin container:

```
 bash-5.1# echo 'p:p_nf_nat_packet_security nf_nat_packet' > /sys/kernel/debug/tracing/kprobe_events
```

This succeeds silently on 1.16.1 but fails on 1.17.0 with the same error message: `No such file or directory`. I've also noticed that this succeeds on `1.17.0` for the `aws-k8s-1.27` variant, which means that this is probably an issue with our latest `kernel-6.1`.

It's not clear to me why this kernel function can no longer be probed. I can see that the symbol exists by running:

```bash
$ grep nf_nat_packet /proc/kallsyms
...
ffff800000f0c140 t nf_nat_packet	[nf_nat]
```

I'll keep looking around and ask folks who are more familiar with Bottlerocket's kernel.

**Comment by markusboehme** [comment]
Thanks for the initial work here, @cbgbt! I poked at this a bit and found the culprit. In case the to be probed function resides in a loadable module, its name needs to be fully qualified (e.g. `nf_nat:nf_nat_packet` since in our configuration the function is loaded from the `nf_nat` module). This is a regression introduced into the kernel in October (https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=b022f0c7e404887a7c5229788fc99eff9f9a80d5).

The regression has been fixed upstream for the latest kernels, and the backports are finding their way to the stable series at the moment. For the 6.1 series, the fix will be in 6.1.71 (https://lore.kernel.org/all/20240103164911.126685591@linuxfoundation.org/) This will be too late for our Bottlerocket 1.18.0 release, but I'll send a PR cherry-picking this one.
